File: testers/avro/java-spring/generated-and-checked-in/com/example/events/CustomerOrderPublisher.java
package com.example.events;

import java.util.concurrent.CompletableFuture;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Service;

/** Type-safe event publisher for customer-order topic */
@Service
public record CustomerOrderPublisher(
    KafkaTemplate<String, CustomerOrder> kafkaTemplate, String topic) {
  public CustomerOrderPublisher(KafkaTemplate<String, CustomerOrder> kafkaTemplate) {
    this(kafkaTemplate, "customer-order");
  }

  public CustomerOrderPublisher withKafkaTemplate(
      KafkaTemplate<String, CustomerOrder> kafkaTemplate) {
    return new CustomerOrderPublisher(kafkaTemplate, topic);
  }

  public CustomerOrderPublisher withTopic(String topic) {
    return new CustomerOrderPublisher(kafkaTemplate, topic);
  }

  /** Publish a CustomerOrder event */
  public CompletableFuture<SendResult<String, CustomerOrder>> publish(
      String key, CustomerOrder event) {
    return kafkaTemplate.send(topic, key, event);
  }
}

File: testers/avro/scala-cats/generated-and-checked-in/com/example/events/CustomerOrderPublisher.scala
package com.example.events

import cats.effect.IO
import fs2.kafka.KafkaProducer
import fs2.kafka.ProducerRecord
import fs2.kafka.ProducerRecords
import fs2.kafka.ProducerResult
import scala.annotation.unused

@unused
/** Type-safe event publisher for customer-order topic */
case class CustomerOrderPublisher(
  kafkaTemplate: KafkaProducer[IO, String, CustomerOrder],
  topic: String = "customer-order"
) {
  /** Publish a CustomerOrder event */
  def publish(
    key: String,
    event: CustomerOrder
  ): IO[ProducerResult[String, CustomerOrder]] = {
    return kafkaTemplate.produce(ProducerRecords.one(ProducerRecord(topic, key, event))).flatten
  }
}